core/src/vaults.ts

```diff
@@ -0,0 +1,378 @@
+        -WAD_NUMBER_OF_DIGITS
+    );
+
+    return {
+        currentUnitPrice: currentUnitCollateralPrice,
+        nextUnitPrice: nextUnitCollateralPrice,
+        nextPriceChange: new Date((lastPriceUpdateTimestampInSeconds + priceUpdateFrequencyInSeconds) * 1000),
+    };
+};
+
+export const getOsmPrices = memoizee(_getOsmPrices, {
+    maxAge: CACHE_EXPIRY_MS,
+    promise: true,
+    length: 2,
+});
+
+const _fetchLiquidationRatio = async (network: string, collateralType: CollateralType) => {
+    const contract = await getContract(network, 'MCD_SPOT');
+    const collateralTypeAsHex = ethers.utils.formatBytes32String(collateralType);
+    const oracleAndLiquidationRatio = await contract.ilks(collateralTypeAsHex);
+    const liquidationRatioAsHex = oracleAndLiquidationRatio.mat;
+    const oracleAddress = oracleAndLiquidationRatio.pip;
+    const liquidationRatio = new BigNumber(liquidationRatioAsHex._hex).shiftedBy(-RAY_NUMBER_OF_DIGITS).toNumber();
+    return { oracleAddress, liquidationRatio };
+};
+
+export const fetchLiquidationRatio = memoizee(_fetchLiquidationRatio, {
+    maxAge: CACHE_EXPIRY_MS,
+    promise: true,
+    length: 2,
+});
+
+const _fetchVaultLiquidationIncentive = async (
+    network: string,
+    liquidatiorContractAddress: string,
+    debtDai: BigNumber
+) => {
+    const liquidatorContractInterface = await getContractInterfaceByName('MCD_CLIP_');
+    const provider = await getProvider(network);
+    const liquidatorContract = new ethers.Contract(liquidatiorContractAddress, liquidatorContractInterface, provider);
+    const incentiveConstantDaiHex = await liquidatorContract.tip();
+    const incentiveRelativeDaiHex = await liquidatorContract.chip();
+    const incentiveRelativeDai = new BigNumber(incentiveRelativeDaiHex._hex)
+        .shiftedBy(-WAD_NUMBER_OF_DIGITS)
+        .multipliedBy(debtDai);
+    const incentiveConstantDai = new BigNumber(incentiveConstantDaiHex._hex).shiftedBy(-RAD_NUMBER_OF_DIGITS);
+    const incentiveCombinedDai = incentiveRelativeDai.plus(incentiveConstantDai);
+    return {
+        incentiveCombinedDai,
+        incentiveConstantDai,
+        incentiveRelativeDai,
+    };
+};
+
+export const fetchVaultLiquidationIncentive = memoizee(_fetchVaultLiquidationIncentive, {
+    maxAge: CACHE_EXPIRY_MS,
+    promise: true,
+    length: 3,
+});
+
+const _isVaultLiquidated = async (network: string, vault: Vault) => {
+    const contract = await getContract(network, 'MCD_DOG');
+    const typeHex = ethers.utils.formatBytes32String(vault.collateralType);
+    const eventFilter = contract.filters.Bark(typeHex, vault.address, null, null, null, null, null);
+    const liquidationEvents = await contract.queryFilter(eventFilter);
+    if (liquidationEvents.length !== 0 && vault.initialDebtDai.eq(0)) {
+        // there was a liquidation and the vault was not used again
+        const latestEvent = liquidationEvents[liquidationEvents.length - 1];
```

Comment:
Btw, what happens if the vault was partially liquidated? Would there be two barking events? Should we then turn this data into an array of objects?

Reply:
if vault is partially liquidated, its debt is not 0. there's the check in the if statement for this.

Dilemma here: we have 2 states - which one should be assigned when there're 3 outcomes?

- I did not see logic in the contracts that forbids the owner from adding collateral and generating dai using the previously liquidated vault.
  - therefore i assume that presence of the event does not necessarily guarantee that that the vault is liquidated.
- Vault can be partially liquidated:
  - event is present, 
  - debt is non-0
  - this does not differ from the previous point where the vault is liquidated and re-used afterwards

Due to the 2 points above, i do not see how these cases can be distinguished and thus i did not add the third `partially-liquidated` state. Instead i assign the vault `not-liquidated` state.

In the ui after the partial liquidation the vault will still is present but will have `not-liquidatable` state due to the limits being hit.

We can distinguish wether the liquidation has been successful by the success of the liquidation function, then we can have the process that searches for the bark event with the specified auction id to fetch relevant to the final ui panel information. Shall i implement this?

Reply:
>  Otherwise when and why do we label the vault as liquidated if their state is indistinguishable from the live ones?

`liquidated` is set if the debt is `0`:

  - in case it was liquidated: the debt is zero and we are correct in stating that is liquidated
  - in case it was liquidated once, then reused somehow and after sometime the debt is 0 due to the owner's actions: we claim it's liquidated. This is not entirely true but makes no difference as far as i see it.